lib/http_client.rb

```diff
@@ -0,0 +1,30 @@
+require 'net/http'
+require 'json'
+
+class HTTPClient
+
+  def get_json(uri:)
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    request = Net::HTTP::Get.new(uri.request_uri)
+
+    response = http.request(request)
+
+    return nil unless response.kind_of? Net::HTTPSuccess
+
+    JSON.parse(response.body)
+  end
+
+  def delete_json(uri:)
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    request = Net::HTTP::Delete.new(uri.request_uri)
+    request.basic_auth @user, @key
```

Comment:
I don't see those properties defined in `HTTPClient`? Those are defined in `BintrayClient` (but `BintrayClient` doesn't inherit `HTTPClient` anyway and uses composition instead – which I think was a good choice – so those are not gonna be available here).

You should probably pass `user` and `key` as parameters to `delete_json` instead – probably as an optional hash which only calls `basic_auth` if that hash is defined, to make it even more flexible

```suggestion
    request.basic_auth auth[:user], auth[:key] unless auth.nil?
```

---

Once you get that in, you could even consider going further with the refactoring, something like this perhaps?

```ruby
  # @param [Net::HTTP class] req_klass
  #        The Net::HTTP class to use to create the request, e.g. Net::HTTP::Get or Net::HTTP::Delete
  # @param [URI] uri The URI to sent the request to
  # @param [Hash] auth If not nil, should be a hash with keys `:user` and `key` that will be used for Basic Auth
  # @return [Hash|Array] The parsed JSON response, or nil if the request was not a HTTPSuccess.
  #
  def json_request(req_klass, uri:, auth: nil)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = (uri.scheme == 'https')
    request = req_klass.new(uri.request_uri)
    request.basic_auth auth[:user], auth[:key] unless auth.nil?

    response = http.request(request)

    return nil unless response.kind_of? Net::HTTPSuccess

    JSON.parse(response.body)
  end

  def get_json(uri:, auth: nil)
    json_request(Net::HTTP::Get, uri: uri, auth: auth)
  end

  def delete_json(uri:, auth: nil)
    json_request(Net::HTTP::Delete, uri: uri, auth: auth)
  end
```

Reply:
Great catch! I thought my unit tests were comprehensive and only run the refactored code in dry run mode. I had overlooked this.

I now have a `.with(basic_auth: ['user', 'key'])` expectation in the test to delete the package which fails with the version of the code on which this comment was made.

I like the DRY suggestion to extract the class, I was looking for something like that but hadn't found it yet.